### ISSUE TITLE
How to write the result from detect.py in text format?

### ISSUE THREAD

**Issue by nichaom** [ultralytics/yolov5#5426]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hi, I would like to write the result from detect.py in text format and reply in Line Application. But the label shows only one class in text format. (Actually, there are 3 classes in the image.) Do you have any suggestion for solving my problem? 
Thank you.

**This is the part of my codes.**

# Print results
                for c in det[:, -1].unique():
                    n = (det[:, -1] == c).sum()  # detections per class
                    s += f'{n} {names[int(c)]}s, '  # add to string

                # Write results
                for *xyxy, conf, cls in reversed(det):
                    if save_txt:  # Write to file
                        xywh = (xyxy2xywh(torch.tensor(xyxy).view(1, 4)) / gn).view(-1).tolist()  # normalized xywh
                        line = (cls, *xywh,conf) if save_conf else (cls, *xywh)  # label format

                    if save_img or view_img:  # Add bbox to image
                        percentage = "{:.0%}".format(conf)
                        label = f'{names[int(cls)]} ({percentage})'
                        plot_one_box(xyxy, im0, label=label, color=colors[int(cls)], line_thickness=3)

            # Save results (image with detections)
            if save_img:
                if dataset.mode == 'image':
                    cv2.imwrite(save_path, im0)

            # Print time (inference + NMS)
            print(f'{s}Done. ({t2 - t1:.3f}s)')

url = request.url_root + '/static/tmp/' + dist_name

  line_bot_api.reply_message(
            event.reply_token, [
                TextSendMessage(text='Object Detection Result: '),
                TextSendMessage(text=label),
                ImageSendMessage(url,url)
            ])

**Comment by glenn-jocher** [comment]
@nichaom I don't quite understand your question, but you should use detect.py to get started, verifying default functionality first and then updating as necessary for your custom workflow. The detections loop and the class of each detection are in the code you pasted above.

https://github.com/ultralytics/yolov5/blob/5d4258fac5e6ceaa9c897f841cb737c56717a996/detect.py#L214-L228

**Comment by nichaom** [comment]
My problem is the text that reply to client is show only one class although there are 2 classes in the image.

**Comment by glenn-jocher** [comment]
@nichaom We don't provide assistance debugging custom code, but detect.py already incorporates single-line output of multiple objects. 
<img width="1068" alt="Screenshot 2021-11-01 at 12 26 40" src="https://user-images.githubusercontent.com/26833433/139664479-f0d2f43d-646d-48c1-bc0b-d0f63cafa036.png">

**Comment by nichaom** [comment]
@glenn-jocher  Thank you.

**Comment by nichaom** [comment]
@glenn-jocher Thank you, I've got some idea. And what's about the %confidence? How can I know %confidence of each object in detect.py? It is showed only amount of object and name of class like 4 persons, 1 bus, 1 fire hydrant.

**Comment by glenn-jocher** [comment]
@nichaom confidences are saved in image annotations and in text file format with --save-txt. You can also use [YOLOv5 PyTorch Hub](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading) inference to return confidences:

```python
import cv2
import torch
from PIL import Image

# Model
model = torch.hub.load('ultralytics/yolov5', 'yolov5s')

# Images
for f in ['zidane.jpg', 'bus.jpg']:
    torch.hub.download_url_to_file('https://ultralytics.com/images/' + f, f)  # download 2 images
img1 = Image.open('zidane.jpg')  # PIL image
img2 = cv2.imread('bus.jpg')[..., ::-1]  # OpenCV image (BGR to RGB)
imgs = [img1, img2]  # batch of images

# Inference
results = model(imgs, size=640)  # includes NMS

# Results
results.print()  
results.save()  # or .show()

results.xyxy[0]  # img1 predictions (tensor)
results.pandas().xyxy[0]  # img1 predictions (pandas)
#      xmin    ymin    xmax   ymax  confidence  class    name
# 0  749.50   43.50  1148.0  704.5    0.874023      0  person
# 1  433.50  433.50   517.5  714.5    0.687988     27     tie
# 2  114.75  195.75  1095.0  708.0    0.624512      0  person
# 3  986.00  304.00  1028.0  420.0    0.286865     27     tie

```

**Comment by sajjad-haider-khan** [comment]
[video_frame_3.txt](https://github.com/ultralytics/yolov5/files/9428377/video_frame_3.txt)

I have this text file as one of the results for the labels from detect.py. I know that 1st column is related to the class but how can I know which these other columns are relating to. As these values are normalized, it is difficult to understand which columns are for xmin, ymin, xmax, ymax, confidence.

Also, I am little confused here because yolov5 accepts (xcentre, y centre, width, height) as input values of bounding boxes but as per your reply above, it outputs values in the form of (xmin, ymin, xmax, ymax).

**Comment by glenn-jocher** [comment]
@ilmseeker --save-txt will save text files in the default YOLOv5 format.


### 1.2 Create Labels

After using a tool like [Roboflow Annotate](https://roboflow.com/annotate?ref=ultralytics) to label your images, export your labels to **YOLO format**, with one `*.txt` file per image (if no objects in image, no `*.txt` file is required). The `*.txt` file specifications are:

- One row per object
- Each row is `class x_center y_center width height` format.
- Box coordinates must be in **normalized xywh** format (from 0 - 1). If your boxes are in pixels, divide `x_center` and `width` by image width, and `y_center` and `height` by image height.
- Class numbers are zero-indexed (start from 0).

<p align="center"><img width="750" src="https://user-images.githubusercontent.com/26833433/91506361-c7965000-e886-11ea-8291-c72b98c25eec.jpg"></p>

The label file corresponding to the above image contains 2 persons (class `0`) and a tie (class `27`):

<p align="center"><img width="428" src="https://user-images.githubusercontent.com/26833433/112467037-d2568c00-8d66-11eb-8796-55402ac0d62f.png"></p>

**Comment by ilmseeker** [comment]
@glenn-jocher thanks for sharing. It's really helpful. However, I was able to figure out the solution from @barney2074's reply in this issue  [https://github.com/ultralytics/yolov5/issues/9160#issuecomment-1227872960]

**Comment by bcugola** [comment]
I'm using yolov8 to do this object detection in video, and I also wanted to store the results in a txt, but what is being stored in the txt is just that the detection has started, the detection data itself is not stored and saved

**Comment by glenn-jocher** [comment]
@bcugola hi there,

To store the detection results in a text format, you can make use of the `--save-txt` argument in the detect.py script. By including this argument, YOLOv5 will save the detection results in a text file in the default YOLO format. Each line of the text file represents a detected object and contains the class index along with the normalized coordinates of the bounding box.

Here's an example command to run the detect.py script and save the results in a text file:
`python detect.py --source path/to/video.mp4 --save-txt`

Please note that the text file will be saved in the same directory as the video file, and it will have the same name as the video file but with a `.txt` extension.

I hope this helps! Let me know if you have any further questions.
